Support CTE/JOIN usage for PIVOT operator

RIC06X opened the pull request:
### Description

This commit fixed an issue that CTE and JOIN cannot be used with pivot operator.

We added 1 field (`Node *context`) in FuncExpr Struct to store pivot additional information. FuncExpr will carry the pivot information to ExecMakeTableFunctionResult. In this function, we pass the pivot information into two newly added FunctionCallInfo that we can retrieve at bbf_pivot function.  

### Issues Resolved

Task: BABEL-4713, BABEL-4714, BABEL-4795

### Test Scenarios Covered ###


### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

RIC06X committed:
[BABEL-4673] Support View usage for PIVOT operator

This commit fixed an issue that view cannot be created or used by stmt
with pivot operator.

Task: BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

deepakshi-mittal commented on test/JDBC/input/pivot-vu-verify.sql:
```diff
@@ -350,6 +336,11 @@ PIVOT (
 
+
+-- Test view of a stmt with pivot operator
+SELECT * FROM pivot_view
```
We generally avoid using select * and also use order by some id to make sure resuts of the tests are consistent always

RIC06X committed:
Add CTE/VIEW/JOIN support for PIVOT

Modified internal PIVOT information analyzer -> executor passing design.
Supported CTE/VIEW/JOIN as well as function/procedure for PIVOT
operator.

Task: BABEL-4713, BABEL-4714, BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
Add VIEW/CTE/JOIN test cases for PIVOT

Added VIEW/CTE/JOIN test cases for pivot, updated schedule files

Task: BABEL-4713, BABEL-4714, BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
[BABEL-4673] Disable CREATE VIEW ON stmt with PIVOT

Disable CREATE VIEW ON stmt with PIVOT operator due to pg_dump cannot
dump pivot view with correct dependency and newly added fields in
FuncExpr casued readFunc failed to parse previous minor version string
parsetree during select from view after mVU.

Task: BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
[BABEL-4673] Remove newly added pivot fields in fcinfo

Removed newly added pivot fields in fcinfo and set the fcinfo->context
to pivot data when the FuncExpr is a bbf_pivot function

Task: BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

xhfanhe commented on test/JDBC/expected/pivot-before-15_7-16_3-vu-verify.out:
```diff
@@ -0,0 +1,1582 @@
+
+
+-- 3 column in src table pivot
+SELECT EmployeeID, [2] AS STORE2, [3] AS STORE3, [4] AS STORE4, [5] AS STORE5, [6] AS STORE6
```
Please add ORDER BY for most of the test cases

RIC06X replied:
Added ORDER BY to all test cases

xhfanhe commented on test/JDBC/expected/pivot-vu-verify.out:
```diff
@@ -1146,6 +1642,16 @@ int#!#int#!#int#!#int#!#int#!#int
 ~~END~~
 
 
+-- Test view of a stmt with pivot operator
```
Please add a line here stating that view with pivot is not yet supported

RIC06X replied:
Added comment of "create view with pivot is not yet supported"

xhfanhe commented on test/JDBC/expected/pivot-vu-verify.out:
```diff
@@ -1124,7 +1233,394 @@ int#!#varchar#!#varchar#!#varchar
+~~END~~
+
+
+-- COMMA JOIN TEST
```
If you are testing all join types here - we also have CROSS JOIN

RIC06X replied:
Added CROSS JOIN tests (with or without CROSS JOIN keyword) and verified results

RIC06X committed:
[BABEL-4713] Add CROSS JOIN test

Added CROSS JOIN test cases.
Updated pivot data save/load logic due to FuncCall fields changes

Task: BABEL-4713, BABEL-4714
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
[BABEL-4713] Move code to Babelfish hook function

Move executor pivot data passing code to Babelfish hook function.

Task: BABEL-4713, BABEL-4714
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -4755,7 +4756,7 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
 	value_col = list_nth_node(ColumnRef, ((FuncCall *)((ResTarget *)stmt->aggFunc)->val)->args, 0);
 	funcName = list_nth_node(String, ((FuncCall *)((ResTarget *)stmt->aggFunc)->val)->funcname, 0);
-	value_colstr = list_nth_node(String, value_col->fields, 0)->sval;
+	value_colstr = list_nth_node(String, value_col->fields, ((List *)value_col->fields)->length - 1)->sval;
```
```
SELECT TOP 5 ManufactureID, [2] AS STORE2, [3] AS STORE3, [4] AS STORE4, [5] AS STORE5, [6] AS STORE6
FROM
(
    SELECT ManufactureID, ItemID, StoreID
    FROM StoreReceipt
)AS srctable
PIVOT (
    COUNT (srctable.ItemID)
    FOR StoreID IN ([2], [3], [4], [5], [6])
) AS pvt
ORDER BY 1
GO
```
This is to handle column name with indirection (srctable.StoreID)

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -4877,16 +4854,41 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+									list_make2(makeString(pstrdup(pstate->p_sourcetext)),
+												makeString(pstrdup(funcName->sval))
+												)
+									);
```
We decided to not to introduce a new BbfPivotData node because it will cause unexpected test failures. Due to time limit, we choose to use list to store pivot data. The first string list is the identifier. We could possibly use ExtensibleNode in the next release which will limit the code change only to the extension.

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -4455,26 +4454,34 @@ bbf_pivot(PG_FUNCTION_ARGS)
+			(errcode(ERRCODE_CHECK_VIOLATION),
+				errmsg("Babelfish PIVOT is not properly initialized.")));
+	}
+	pivot_parsetree = (List *) list_nth((List *) fcinfo->context, 1);
```
Since we're reusing fcinfo->context, pls add some test case to : 
1. trigger, inside trigger procedure run a pivot query
2. table function, in the def of the function, pls run a pivot query

RIC06X replied:
Thanks for the comments! 
1. Will add a trigger test. 
2. We already have a table function test cases which is define in `pivot-vu-prepare.sql:433`.

https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/45d60a99620d323660f4e3f77578de6ad3fb43d1/test/JDBC/input/pivot-vu-prepare.sql#L443

RIC06X committed:
[BABEL-4713] Add trigger test for pivot

Add trigger test for pivot. Removed unnecessary pivot test scripts.

Task: BABEL-4713, BABEL-4714
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>
